### DISCUSSION TITLE
Issue writing text responses to data

### DISCUSSION THREAD

**Discussion by jwylie21** [jspsych/jsPsych#2011] [Q&A] [answered]
Hello!

I am conducting an experiment where participants will write an opinion about a current issue and then write in words they see during an RSVP task. Everything is working great except the write in data show up as [object Object]. I'm not really sure what I'm missing here to get the actual keyed responses recorded. 

Here is the code for the write-in: 

```
var manip2 = {
	type: 'survey-text',
	questions: [
	  {prompt: "In early 2020 the COVID-19 pandemic swept across the world. Do you think it would be appropriate for the United States to make receiving the COVID-19 vaccine mandatory? Why or why not?</p>", 
	  rows: 6, columns: 40}
	],
  };  
```

Here is the code for the RSVP: 

```
var response_block = {
  type: 'survey-text',
  questions: [{prompt: "Which words appeared in green? Please separate them with a comma."}],
  data: jsPsych.timelineVariable('data'),
}

function make_rsvp_timeline(trials, phase) {
	rsvp_timeline = [];
	trial_number  = 0;
	for (trial in trials) {
		trial_number++;
		rsvp_stimuli = rsvp_trial(trials[trial]);
                var rsvp_block_stimuli = [];
		for (stimulus in rsvp_stimuli.stimuli) {
			rsvp_block_stimuli.push(
	  			{
	  				stimulus: "<span class='rsvp'>" + rsvp_stimuli.stimuli[stimulus] + "</span>",
	  				data: {
	  					phase: phase,
	  					test_part: 'rsvp',
	  					stim: rsvp_stimuli.stimuli[stimulus],
	  					trial_number: trial_number
	  				}
	  			}
	  		);
		}
		if (phase == 'instructions') {
			// slow stimuli
			stimulus_trial = rsvp_demo_stimulus_block;
			iti_trial      = rsvp_demo_iti;
		} else {
			stimulus_trial = rsvp_stimulus_block;
			iti_trial      = rsvp_iti;
		}
		var test_procedure = {
			timeline: [stimulus_trial, iti_trial],
			timeline_variables: rsvp_block_stimuli
		}
	  	var rsvp_response_stimuli = [];  // Getting a responses
	  	rsvp_response_stimuli.push(
			{
				data: {
					phase: phase,
					test_part: 'response',
					trial_number: trial_number
				}
			}
		);

var response_procedure = {
			timeline: [response_block],
			timeline_variables: rsvp_response_stimuli
		}
```

Any ideas would be greatly appreciated!

**Comment by becky-gilbert** [comment]
Hi @jwylie21, this code is a little hard to follow - it's hard to understand what's going on without more information. For instance, the `make_rsvp_timeline` function uses variables such as `trials` and `rsvp_stimuli` and the function `rsvp_trial`, but I don't know what those things look like. Maybe you could post a simplified example that contains all of the code necessary to generate 1-2 trials? Or perhaps you could provide some code that demonstrates the problem using trials that are generated 'manually' rather than via a function? Also feel free to post a link to an online version, since that would be an easy way to share all of your code. Thanks!

**Reply by jwylie21** [reply]
Hi @becky-gilbert! [Here](https://gitlab.pavlovia.org/nichdima/moral_blinding/blob/master/experiment.js) is a link to the code on GitLab.

Let me know if that doesn't work for some reason.

**Reply by becky-gilbert** [reply]
Hi @jwylie21, thanks for sharing your code - very helpful! 
Can you please clarify what you mean when you say that "the write in data show up as [object Object]"? Do you mean that a _stimulus_ appears on the screen as "[object Object]", and if so, which one? Or do you mean that the participant's responses to one of the survey-text trials (`manip2`, `response_block`) show up as "[object Object]" in the _data_? Thanks!

**Comment by jwylie21** [comment]
Hi @becky-gilbert! It is the last part! When I look at the data file the type participant responses, they show up like "[object Object]". Attaching a datafile example here: 

[moral_blinding_PARTICIPANT_SESSION_2021-07-12_15h27.59.836.csv](https://github.com/jspsych/jsPsych/files/6910135/moral_blinding_PARTICIPANT_SESSION_2021-07-12_15h27.59.836.csv)

**Comment by becky-gilbert** [comment, marked as answer]
Hi @jwylie21, I think the problem is that you're using v6.1.0 for the core jspsych.js file:
```html
<script type="text/javascript" src="lib/vendors/jspsych-6.1.0/jspsych.js"></script>
```
But it looks like the plugin files that you're using come from the jsPsych v6.3.0 release. There were a number of changes between these releases - [see here for release notes](https://github.com/jspsych/jsPsych/releases). One of the changes was that the jsPsych data switched from internally storing complex data (objects and array) as JSON strings to storing these as native JavaScript objects/arrays, and then converting these JavaScript variables to strings in the last step of data saving (i.e. jsPsych.data.get().csv()). What this means in your case is that the plugins are saving the data as _JavaScript_ objects/arrays, but because your jsPsych core file is older, it isn't converting these into strings.

I think you have two options for dealing with this. One option is to switch to using the jspsych.js file from v6.3.0, so that it's consistent with the plugin versions that you're using. This is the best option IMO, but you should be aware that switching your jspsych.js file might cause other compatibility problems that you'll then have to fix. For instance, some of the plugin parameter and data names have changed, preloading moved from `jsPsych.init` into a plugin, and jsPsych switched from using numeric key codes (e.g. `choices: [74]`) to key strings (e.g. `choices: ['j']`).

If you don't want to have to fix these other compatibility issues, then another option is to convert any complex data from JavaScript to JSON/CSV strings in the trial's `on_finish` function. Converting to JSON is straightforward:
```javascript
  var manip2 = {
    type: 'survey-text',
    questions: [
      {prompt: "In early 2020 the COVID-19 pandemic swept across the world, resulting in significant loss of life. To make matters worse, many Americans have lost their jobs and livelihoods as a result of the pandemic. To slow the spread of the virus, the CDC has recommended that people get the COVID-19 vaccine. Some locations have made getting the vaccine mandatory. Despite the risk of spreading the virus, many people still choose not to get the COVID-19 vaccine. <p>In your own words, do you think it would be appropriate for the United States to make receiving the COVID-19 vaccine mandatory? Why or why not?</p>", rows: 6, columns: 40}
    ],
    on_finish: function(data) {
      data.response_json = JSON.stringify(data.response);
    }
  };
```
I _think_ once you do that, the data should then be converted to CSV when the data is sent back to Pavlovia. But if not, you might end up with a JSON-formatted string inside the `response_json` variable that you then have to parse.

**Reply by jwylie21** [reply]
Hi @becky-gilbert! Thank you sooooooo much! I am using the on_finish solution for this experiment (but will move to v6.3.0 in the future. 

I have one more question about getting the data written (vs. [object, Object]). 

I have added an on_finish solution to everything, but am struggling to get condition written in the data even though I have it listed as part of: 

```
jsPsych.data.addProperties({
	subject: subject_id,
	survey: surveyCode,
	condition: cond,
	brwser: browser,
	operatingS: OS,
 block_order: blockorder
 });
```

I'm assuming that condition is not working (but all the others are) because it is a procedure vs. a variable. Here is the part of the code that defines the condition procedure: 

```
var cond_procedure1 = {
timeline: [manip1], 
randomize_order: false 
};

var cond_procedure2 = {
  timeline: [manip2],
  randomize_order: false
};

var conditions = [cond_procedure1, cond_procedure2];
var cond = jsPsych.randomization.sampleWithoutReplacement(conditions,1)[0];
console.log('cond: ', +cond);  // check console to see which condition was selected
```
Any ideas?

**Reply by becky-gilbert** [reply]
Yes, you're right that `cond` isn't saving to the data because it's an object. You could use `JSON.stringify` to convert it from an object to a string, but that's probably not ideal since I think what you're interested in is just the condition number, not the whole procedure object. So I would do it like this:
```js
var cond_procedure1 = {
  timeline: [manip1], 
  randomize_order: false 
};

var cond_procedure2 = {
  timeline: [manip2],
  randomize_order: false
};

// randomly select a condition number: 1 or 2
var cond = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
console.log('cond: ', cond);  // check console to see which condition number was selected

// use the condition number to get the appropriate procedure object 
// and add it to the main timeline (or whatever you need to do with it)
if (cond == 1) {
  timeline.push(cond_procedure1);
} else if (cond == 2) {
  timeline.push(cond_procedure1);
}

jsPsych.data.addProperties({
  // other properties...
  condition: cond, // save the condition number to the data
});
```

**Reply by jwylie21** [reply]
Amazing, thank you SO much!

**Reply by jwylie21** [reply]
Hi @becky-gilbert , 

I have come across the issue again! (Next experiment, I will certainly move to 6.3). 

I'm wondering if you have any ideas for recording IP Addresses. I can see that the IP Address is being recorded to the console again, but it shows up as [object Object] on the data side. However, because it isn't a jspsych kind of thing (the $.getJSON), I'm not sure how to access that information. I've tried a million things and am out of ideas. Any suggestions would be appreciated!! 

```
var ip = $.getJSON("https://api.ipify.org?format=jsonp&callback=?", function(data) {
	data = console.log(JSON.stringify(data, null, 2))
});
jsPsych.data.addProperties({IP: ip}) 
```
